Pitch: String processing algorithms (#188)
apple/swift-experimental-string-processing: Documentation/Evolution/StringProcessingAlgorithms.md

    public func ranges<R: RegexProtocol>(of regex: R) -> some Collection<Range<Index>>
}
```

### Use custom parsers in regex builders and `RegexProtocol` algorithms

We want to extend string processing to types from outside the standard library, so that one can use custom parsers with the string processing algorithms and incorporate custom parsers in regex builders seamlessly. 

Consider parsing an HTTP header to capture the date field as a `Date` type:

```swift
let header = """
HTTP/1.1 301 Redirect
Date: Wed, 16 Feb 2022 23:53:19 GMT
Connection: close
Location: https://www.apple.com/
"""
```

You are likely going to match a substring that look like a date string (`16 Feb 2022`), and parse the substring as a `Date` with one of the date parsers in the Foundation framework:

```swift
let regex = Regex {
    capture {
        oneOrMore(.digit)
        " "
        oneOrMore(.word)
        " "
        oneOrMore(.digit)
    }
}

let dateParser = Date.ParseStrategy(format: "\(day: .twoDigits) \(month: .abbreviated) \(year: .padded(4))"
if let dateMatch = header.firstMatch(of: regex)?.0 {
    let date = try? Date(dateMatch, strategy: dateParser)
}
```

> This works, but wouldn't it be much more approachable if you can directly use the date parser within the string match function?

Reviewer: ```suggestion
While this approach happens to work for this example, it is fragile and more onerous than just directly using Foundation's date parser directly within the regex.
```

I would recommend against rhetorical questions, especially "wouldn't it be better" style ones.

Reviewer: Also, you can consider streamlining some of this as part of an extended motivation section.

```swift
let regex = Regex {
    capture(dateParser)
}

let date = header.firstMatch(of: regex).map(\.result.1) 
// A `Date` representing 2022-02-16 00:00:00 +0000
```

Or consider parsing a bank statement to record all the monetary values in the last column:

```swift
let statement = """
CREDIT    04/06/2020    Paypal transfer    $4.99
CREDIT    04/03/2020    Payroll            $69.73
DEBIT     04/02/2020    ACH transfer       ($38.25)
DEBIT     03/24/2020    IRX tax payment    ($52,249.98)
"""
```

> We have already seen that parsing a date string can be tricky since it could contain localized month name (`"Feb"` as seen from above). Parsing a currency string such as `$3,020.85` with regex is not trivial either -- it can contain grouping separators, a decimal separator, and a currency symbol, all of which can be localized. 
> 
> Foundation has various parsers for localized strings like these. Delegating this task to dedicated parsers alleviates the need to handle it yourself. In the second part of the pitch, we introduce the `CustomRegexComponent` protocol that conveniently lets types from outside the standard library participate in regex builders and `RegexProtocol` algorithms.

Reviewer: ```suggestion
Parsing a date string can be tricky, as it could contain localized information (`"Feb"` as seen from above). Parsing a currency string such as `$3,020.85` with regex is also tricky -- it can contain grouping separators, a decimal separator, and a currency symbol, all of which can be localized. This is why Foundation provides industrial-strength parsers for localized strings like these.

 We propose a `CustomRegexComponent` protocol which allows types from outside the standard library participate in regex builders and `RegexProtocol` algorithms.
```

My suggestions are minor localized editorial changes. But, we're repeat the same information throughout, so there's more tidying up. Since this is the proposed solution section, it's a little weird to be talking about the "second part of this pitch", because this is where the second part goes.

> If `Foundation.FloatingPointFormatStyle<Double>.Currency` conformed to `CustomRegexComponent`, you would be able to retrieve the currency from the bank statement above as a list of `Double` values this way:

Reviewer: ```suggestion
`CustomRegexComponent` allows types, such as `Foundation.FloatingPointFormatStyle<Double>.Currency`, to be used directly within a regex.
```

I understand that you're not proposing the conformances for Foundation here, but I don't think the "if ... you would be able to..." adds clarity.

    /// - Returns: `true` if the collection contains the specified sequence,
    /// otherwise `false`.
    public func contains<S: Sequence>(_ other: S) -> Bool
        where S.Element == Element
}

extension BidirectionalCollection where SubSequence == Substring {
    /// Returns a Boolean value indicating whether the collection contains the
    /// given regex.
    /// - Parameter regex: A regex to search for within this collection.
    /// - Returns: `true` if the regex was found in the collection, otherwise
    /// `false`.
    public func contains<R: RegexProtocol>(_ regex: R) -> Bool
}
```

#### Starts with

```swift
extension BidirectionalCollection where SubSequence == Substring {
    /// Returns a Boolean value indicating whether the initial elements of the
    /// sequence are the same as the elements in the specified regex.
    /// - Parameter regex: A regex to compare to this sequence.
    /// - Returns: `true` if the initial elements of the sequence matches the
    /// beginning of `regex`; otherwise, `false`.
    public func starts<R: RegexProtocol>(with regex: R) -> Bool
}
```

#### Trim prefix

```swift

> extension Collection {

Author: I find myself squinting to see the differences between these functions. I wonder if it's easier to review if we group them by extension other than algorithms, like

```swift
extension Collection where Element: Equatable {
    public func contains<S: Sequence>(_ other: S) -> Bool
    public func trimmingPrefix<Prefix: Collection>(_ prefix: Prefix) -> SubSequence
}

extension BidirectionalCollection where SubSequence == Substring {
    public func contains<R: RegexProtocol>(_ regex: R) -> Bool
    public func starts<R: RegexProtocol>(with regex: R) -> Bool
}
```

}

extension BidirectionalCollection where SubSequence == Substring {
    public func split<R: RegexProtocol>(by separator: R) -> some Collection<Substring>
}
```

### `CustomRegexComponent` protocol

The `CustomRegexComponent` protocol inherits from `RegexProtocol` and satisfies its sole requirement. This enables the usage of types that conform to `CustomRegexComponent` in regex builders and `RegexProtocol` algorithms.

```swift
public protocol CustomRegexComponent: RegexProtocol {
    /// Match the input string within the specified bounds, beginning at the given index, and return
    /// the end position (upper bound) of the match and the matched instance.
    /// - Parameters:
    ///   - input: The string in which the match is performed.
    ///   - index: An index of `input` at which to begin matching.
    ///   - bounds: The bounds in `input` in which the match is performed.
    /// - Returns: The upper bound where the match terminates and a matched instance, or nil if
    ///   there isn't a match.
    func match(
        _ input: String,
        startingAt index: String.Index,
        in bounds: Range<String.Index>
    ) -> (upperBound: String.Index, match: Match)?
}
```

> You can conform your custom parser to `CustomRegexComponent`. Conformance is simple: implement the `match` function to return the upper bound of the matched substring, and the type represented by the matched range. It inherits from `RegexProtocol`, so you will be able to use it with all of the string algorithms that take a `RegexProtocol` type.

Reviewer: ```suggestion
Custom parsers can conform by implementing the `match` function to return the upper bound of the matched substring, and the value constructed by the match. Conformers naturally inherit from `RegexProtocol`, so they can be used with all of the string algorithms generic over `RegexProtocol`.
```

> Here, we use Foundation framework's `FloatingPointFormatStyle<Double>.Currency` as an example. `FloatingPointFormatStyle<Double>.Currency` would conform to `CustomRegexComponent` by implementing the `match` function, with `Match` being a `Double`. It could also add a static function `.localizedCurrency(code:)` as a member of `RegexProtocol`, so you can refer to it as `.localizedCurrency(code:)` in the `Regex` result builder. 

Reviewer: It really seems like this should be merged into the "proposed solution" sub-section, where the example's fresh in the reader's mind.

```swift
extension FloatingPointFormatStyle<Double>.Currency : CustomRegexComponent { 
    func match(
        _ input: String,
        startingAt index: String.Index,
        in bounds: Range<String.Index>
    ) -> (upperBound: String.Index, match: Double)?
}

extension RegexProtocol where Self == FloatingPointFormatStyle<Double>.Currency {
    public static func localizedCurrency(code: Locale.Currency) -> Self
}
```

> Users could specify a pattern to match a localized currency amount such as `"$3,020.85"` simply with the following, and use it in any of the string matching algorithms introduced above.

Reviewer: ```suggestion
Matching and extracting a localized currency amount, such as `"$3,020.85"`, can be done directly within a regex:
```

    ///   sequence matching `regex` to replace. Default is `Int.max`.
    /// - Returns: A new collection in which all occurrences of subsequence
    /// matching `regex` are replaced by `replacement`.
    public func replacing<R: RegexProtocol, Replacement: Collection>(
        _ regex: R,
        with replacement: Replacement,
        maxReplacements: Int = .max
    ) -> Self where Replacement.Element == Element
  
    /// Replaces all occurrences of the sequence matching the given regex with
    /// a given collection.
    /// - Parameters:
    ///   - regex: A regex describing the sequence to replace.
    ///   - replacement: The new elements to add to the collection.
    ///   - maxReplacements: A number specifying how many occurrences of the
    ///   sequence matching `regex` to replace. Default is `Int.max`.
    public mutating func replace<R: RegexProtocol, Replacement: Collection>(
        _ regex: R,
        with replacement: Replacement,
        maxReplacements: Int = .max
    ) where Replacement.Element == Element
  
    public func replacing<R: RegexProtocol, Replacement: Collection>(
        _ regex: R,

>         with replacement: (RegexMatch<R.Match>) throws -> Replacement,

Author: @timvermeulen @milseman I need help for documentation here. I'm not sure what this closure does.

Reviewer: The closure receives the full match info, including captures, and returns a replacement string.

We propose:

1. New regex-powered algorithms over strings, bringing the standard library up to parity with scripting languages
2. Generic `Collection` equivalents of these algorithms in terms of subsequences
3. `protocol CustomMatchingRegexComponent`, allowing libraries to vend types that can be intermixed as components of regexes

This proposal is part of a larger [regex-powered string processing initiative](https://forums.swift.org/t/declarative-string-processing-overview/52459). Throughout the document, we will reference the still-in-progress [`RegexProtocol`, `Regex`](https://github.com/apple/swift-experimental-string-processing/blob/main/Documentation/Evolution/StronglyTypedCaptures.md), and result builder DSL, but these are in flux and not formally part of this proposal. Further discussion of regex specifics is out of scope of this proposal and better discussed in another thread (see [Pitch and Proposal Status](https://github.com/apple/swift-experimental-string-processing/issues/107) for links to relevant threads).

## Motivation

> Below is how [Python string processing API](https://docs.python.org/3/library/stdtypes.html#text-sequence-type-str) compare to their counterparts in Swift. Many essential string processing APIs are evidently missing in the standard library. While most of those can be substituted by chaining multiple functions or using Foundation, they are basic enough to warrant a place in the standard library.

Reviewer: ```suggestion
Below is a comparison of how Swift's APIs stack up with Python's. Some of the gaps can be accomplished through a series of API calls, but every gap adds a burden to developers doing frequent or complex string processing.
```

Reviewer: Just word smithing. You can put the hyperlink back in too.

| `removeprefix(prefix)` |  |  |
| `removesuffix(suffix)` |  |  |
| `replace(old, new, count)` |  | Alternative: `replacingOccurrences(of:with:)` in Foundation |
| `rfind(sub, start, end)` | `lastIndex(where:)` | Does not allow specifying a search range |
| `rindex(sub, start, end)` | `lastIndex(where:)` | Same as above |
| `rjust(width, fillchar)` |  |  |
| `rpartition(sep)` |  |  |
| `rsplit(sep, maxsplit)` |  |  |
| `rstrip([chars])` |  |  |
| `split(sep, maxsplit)` | `split(separator:maxSplits:...)` |  |
| `splitlines(keepends)` | `split(separator:maxSplits:...)` |  |
| `startswith(prefix, start, end)` | `starts(with:)` |  |  |
| `strip([chars])` |  | Alternative: `trimmingCharacters(in:)` in Foundation  |
| `translate(table)` |  |  |
| `zfill(width)` |  | Alternative: `padding(toLength:)` in Foundation |

Note: The comparison table omits functions to query if all characters in the string are of a specified category, such as `isalnum()` and `isalpha()`. These are achievable in Swift by passing in the corresponding character set to `allSatisfy(_:)`, so they're omitted in this table for simplicity. 

> ### Processing domain-specific strings

Reviewer: ```suggestion
### Complex string processing
```

> Another common task regarding string processing is handling those with domain specific information. 

Reviewer: ```suggestion
Even with these API additions, more complex string processing quickly becomes unwieldy. Up-coming support for authoring regexes in Swift help alleviate this somewhat, but string processing in the modern world involves dealing with localization, standards-conforming validation, and other concerns for which a dedicated parser is required.
```

Consider parsing the date field `"Date: Wed, 16 Feb 2022 23:53:19 GMT"` in an HTTP header as a `Date` type. The naive approach is to search for a substring that looks like a date string (`16 Feb 2022`), and attempt to post-process it as a `Date` with a date parser:

```swift
let regex = Regex {
    capture {
        oneOrMore(.digit)
        " "
        oneOrMore(.word)
        " "
        oneOrMore(.digit)
    }
}

let dateParser = Date.ParseStrategy(format: "\(day: .twoDigits) \(month: .abbreviated) \(year: .padded(4))"
if let dateMatch = header.firstMatch(of: regex)?.0 {
    let date = try? Date(dateMatch, strategy: dateParser)
}
```

> While this approach happens to work for this example, it is fragile when it comes to localized strings. 

Reviewer: ```suggestion
This requires writing a simplistic pre-parser before invoking the real parser. The pre-parser will suffer from being out-of-sync and less featureful than what the real parser can do.
```

Or consider parsing a bank statement to record all the monetary values in the last column:

```swift
let statement = """
CREDIT    04/06/2020    Paypal transfer    $4.99
CREDIT    04/03/2020    Payroll            $69.73
DEBIT     04/02/2020    ACH transfer       ($38.25)
DEBIT     03/24/2020    IRX tax payment    ($52,249.98)
"""
```

> Parsing a currency string such as `$3,020.85` with regex is also tricky, as it can contain localized and currency symbols. This is why Foundation provides industrial-strength parsers for localized strings like these. 

Reviewer: ```suggestion
Parsing a currency string such as `$3,020.85` with regex is also tricky, as it can contain localized and currency symbols in addition to accounting conventions. This is why Foundation provides industrial-strength parsers for localized strings.
```

Here, we use Foundation `FloatingPointFormatStyle<Double>.Currency` as an example for protocol conformance. It would implement the `match` function with `Match` being a `Double`. It could also add a static function `.localizedCurrency(code:)` as a member of `RegexComponent`, so it can be referred as `.localizedCurrency(code:)` in the `Regex` result builder:

```swift
extension FloatingPointFormatStyle<Double>.Currency : CustomMatchingRegexComponent { 
    public func match(
        _ input: String,
        startingAt index: String.Index,
        in bounds: Range<String.Index>
    ) -> (upperBound: String.Index, match: Double)?
}

extension RegexComponent where Self == FloatingPointFormatStyle<Double>.Currency {
    public static func localizedCurrency(code: Locale.Currency) -> Self
}
```

Matching and extracting a localized currency amount, such as `"$3,020.85"`, can be done directly within a regex:

```swift
let regex = Regex {
    capture(.localizedCurreny(code: "USD"))
}
```

## Alternatives considered

### Extend `Sequence` instead of `Collection`

> All of the proposed algorithms are specific to the `Collection` protocol, without support for plain `Sequence`s. Types conforming to the `Sequence` protocol are not required to support multi-pass iteration, which makes a `Sequence` conformance insufficient for most of these algorithms. In light of this, the decision was made to have the underlying shared algorithm implementations work exclusively with `Collection`s.

Reviewer: ```suggestion
Most of the proposed algorithms are necessarily on `Collection` due to the use of indices or mutation.  `Sequence` does not support multi-pass iteration, so even `trimPrefix` would problematic on `Sequence` because it needs to look 1 `Element` ahead to know when to stop trimming.
```

# String processing algorithms

## Introduction

The Swift standard library's string processing algorithms are underpowered compared to other popular programming and scripting languages. Some of these omissions can be papered over through bridged `NSString` API available through the Foundation framework, but these are still incomplete and bring in linkage requirements to Foundation.

We propose:

1. New regex-powered algorithms over strings, bringing the standard library up to parity with scripting languages
2. Generic `Collection` equivalents of these algorithms in terms of subsequences
3. `protocol CustomRegexComponent`, allowing libraries to vend types that can be intermixed as components of regexes

> This proposal is part of a larger [regex-powered string processing initiative](https://forums.swift.org/t/declarative-string-processing-overview/52459). Throughout the document, we will reference the still-in-progress [`RegexProtocol`, `Regex`](https://github.com/apple/swift-experimental-string-processing/blob/main/Documentation/Evolution/StronglyTypedCaptures.md), and [result builder DSL](https://forums.swift.org/t/pitch-regular-expression-literals/52820), but these are in flux and not formally part of this proposal. Further discussion of regex specifics is out of scope of this proposal and better discussed in another thread (see [Pitch and Proposal Status](https://github.com/apple/swift-experimental-string-processing/issues/107) for links to relevant threads).

Reviewer 2: The "result builder DSL" is linking to the wrong pitch though, the link points to the regex literal pitch. We don't have a DSL pitch yet — I think you can just use plain text for now, or link to https://github.com/rxwei/swift-experimental-string-processing/blob/dsl-api-walkthrough/Documentation/DSLAPIWalkthough.md (which I plan to turn into a pitch very soon).

Author: Thanks for catching that. I removed the link all together.
